File: external/java/src/main/java/ch/epfl/dedis/lib/omniledger/Spawn.java
package ch.epfl.dedis.lib.omniledger;

import ch.epfl.dedis.proto.OmniLedgerProto;

import java.util.ArrayList;
import java.util.List;

/**
 * Spawn is an operation that an Instruction can take, it should be used for creating an object.
 */
public class Spawn {
    private String contractId;
    private List<Argument> arguments;

    /**
     * Constructor for the spawn action.
     * @param contractId The contract ID.
     * @param arguments The initial arguments for running the contract.
     */
    public Spawn(String contractId, List<Argument> arguments) {
        this.contractId = contractId;
        this.arguments = arguments;
    }

    public Spawn(OmniLedgerProto.Spawn proto) {
        contractId = proto.getContractid();
        arguments = new ArrayList<Argument>();
        for (OmniLedgerProto.Argument a : proto.getArgsList()) {
            arguments.add(new Argument(a));
        }
    }

    /**
     * Getter for contract ID.
     * @return The contract ID.
     */
    public String getContractId() {
        return contractId;
    }

    /**
     * Getter for the arguments.
     * @return The arguments.
     */
    public List<Argument> getArguments() {
        return arguments;
    }

    /**
     * Converts this object to the protobuf representation.
     * @return The protobuf representation.
     */
    public OmniLedgerProto.Spawn toProto() {
        OmniLedgerProto.Spawn.Builder b = OmniLedgerProto.Spawn.newBuilder();
        b.setContractid(this.contractId);
        for (Argument a : this.arguments) {
            b.addArgs(a.toProto());
        }
        return b.build();
    }
}

File: external/js/cothority/lib/omniledger/Spawn.js
/**
 * Spawn is an operation that an Instruction can take, it should be used for creating an object.
 */
class Spawn {
  /**
   * Constructor for the spawn action.

   * @param {string} contractId - The contract ID
   * @param {Argument[]} argumentsList - The initial arguments for running the contract.
   */
  constructor(contractId, argumentsList) {
    this._contractId = contractId;
    this._arguments = argumentsList;
  }

  /**
   * @return {string} - The contract ID
   */
  get contractId() {
    return this._contractId;
  }

  /**
   * @return {Argument[]} - The arguments
   */
  get argumentsList() {
    return this._arguments;
  }

  /**
   * Create an object with all the necessary field needed to be a valid message
   * in the sense of protobufjs. This object can then be used with the "create"
   * method of protobuf
   *
   * @return {Object}
   */
  toProtobufValidMessage() {
    return {
      contractid: this._contractId,
      args: this._arguments.map(arg => arg.toProtobufValidMessage())
    };
  }
}

module.exports = Spawn;
